[nnc] Updated IR cloning to create clones of expressions in addition to statements

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62833

Differential Revision: [D30135980](https://our.internmc.facebook.com/intern/diff/D30135980)

navahgar committed:
[nnc] Updated IR cloning to create clones of expressions in addition to statements

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30135980).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_cloner.cpp:
```diff
@@ -0,0 +1,343 @@
+#include <torch/csrc/jit/tensorexpr/ir_cloner.h>
+
+#include <torch/csrc/jit/tensorexpr/eval.h>
```
Where is this used?

navahgar replied:
Removed.

huiguoo commented on torch/csrc/jit/tensorexpr/ir_cloner.h:
```diff
@@ -0,0 +1,111 @@
+class SyncThreads;
+class ExternalCall;
+
+class TORCH_API IRCloner : public IRMutator {
```
Should we separate IRCloner from IRMutator? They server for very different purposes and it looks weird that we do `clone` in the name of `mutate`. Since the IRCloner has its own mutations on expressions and stmts now, I guess there's not much code it can reuse from IRMutator?

navahgar replied:
Actually, I considered this and I personally didn't have a strong preference either way. And it was just simpler to keep it same as before.

If you have a strong preference to doing this, I can do it in a separate PR.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_cloner.cpp:
```diff
@@ -0,0 +1,343 @@
+    case IRNodeType::kRshift:
+      return new Rshift(lhs_new, rhs_new);
+    default:
+      throw unsupported_dtype();
```
Nit: we probably should throw a different error here, it's not related to dtype.

navahgar replied:
Yeah. Replaced with `unimplemented_lowering`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_cloner.cpp:
```diff
@@ -0,0 +1,343 @@
+  return new Load(v->dtype(), buf_new, indices_new);
+}
+
+Expr* IRCloner::mutate(Buf* v) {
```
That's a bit awkward that we're cloning Buf, but not cloning Vars. I understand why we might have to do this (since Buf might contain expressions in dims that need to be cloned), but it still could be a surprising behavior. Maybe for consistency we then should be cloning Vars too? Or if that's too breaking, we might consider not cloning Bufs (and its inner expressions). Either way, please document the desired behavior somewhere

navahgar replied:
I debated about this. But went ahead with cloning buffers for the exact reasons you mentioned (because the expressions in dims and initializers). 

Since the difference between cloning Buf and not cloning Var could be confusing as you said, I have reverted to not cloning both of them. The only case I can think of so far which could change the expressions in dims of Buf is buffer compression. And for that case, anyways all the kernel code should be analyzed and it is better to have the clones point to the same Buf. 

If something comes up later, we can revisit this assumption and add cloning for Buf and Var.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_cloner.cpp:
```diff
@@ -0,0 +1,343 @@
+    indices_new.push_back(ind->accept_mutator(this));
+  }
+  auto value_new = v->value()->accept_mutator(this);
+  return new Store(v->buf(), indices_new, value_new);
```
We have a mutator for Buf, but we're not calling it here. Is that intentional? In what situations are we cloning the Buf?

navahgar replied:
Yeah, I missed that. Fixed now.

ZolotukhinM reviewed (commented):
Overall looks good, I have one question about cloning behavior with Bufs (see below).

navahgar committed:
Update on "[nnc] Updated IR cloning to create clones of expressions in addition to statements"


Differential Revision: [D30135980](https://our.internmc.facebook.com/intern/diff/D30135980)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30135980).

ZolotukhinM reviewed (approved):
LGTM!

navahgar committed:
Update on "[nnc] Updated IR cloning to create clones of expressions in addition to statements"


Differential Revision: [D30135980](https://our.internmc.facebook.com/intern/diff/D30135980)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30135980).

navahgar committed:
Update on "[nnc] Updated IR cloning to create clones of expressions in addition to statements"


Differential Revision: [D30135980](https://our.internmc.facebook.com/intern/diff/D30135980)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30135980).

navahgar committed:
Update on "[nnc] Updated IR cloning to create clones of expressions in addition to statements"


Differential Revision: [D30135980](https://our.internmc.facebook.com/intern/diff/D30135980)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30135980).
